### ISSUE TITLE
Merge bind and listen?

### ISSUE THREAD

**Issue by sunfishcode** [WebAssembly/WASI#741]
I know this has been iterated on in other contexts, but

Implementing wasi-sockets on top of the Rust standard library, I have to bridge the gap that Rust's `bind` function does `socket`, `bind`, and `listen` all in one step. I have it implemented with a little state machine that collects up the socket and bind info and does everything at the point of the listen. But this does mean that if users pass port 0 to `bind` and then do `getsockname` afterwards to find out what port it picked, I can't implement it because I haven't actually created a socket at that point.

The answer might be that the Rust standard library is too high-level here and I need to use lower-level APIs, and I can do that. But I suspect Rust is not the only language that does something like this. If so, it's worth considering merging these calls.

**Comment by Ericson2314** [comment]
I understand the implementation headaches when changing things, but I also lean towards rethinking the API here. The traditional sockets one is truly atrocious.

(I wrote about this once here https://github.com/Ericson2314/baccumulation/blob/main/reforming-unix.adoc#networking)

Another motivation is trying to reduce ambient authority / increase protocol agnosticism. The current dance support famously anonymous sockets on the server side via socket activation. And this also allows for a server not to care whether it is listening on TCP or UDP --- it just `accept`s on the listening socket it has been given.

On the client side, this is not the case at all. It would be nice to given a capability which represents something that can be connected. This could be network address to dial, or a local unix socket, and just connect to that. `connect` would, just like `accept`, return a new file descriptor, since the interface that potentially supports multiple connections != a specific connection.

To some up:

- listening sockets / dialing sockets --- call accept/connect to get a connection socket
- Network ones: can choose an external address (IP, port, etc.) and make either a listening socket with bind/listen, or a dialing socket with <new thing>
- Local sockets, should create a pair of each, (not like `socket` pair, that creates a pair for a single connection only, not making connections)
- Attaching a local dialing socket to the FS is what `bind` in effect does today. Attaching a listening socket would be the hypothetical opposite thing --- a "file" that grants the opener the ability to `accept` but not `connect`.

The current APIs messiness --- all these artificial state machine steps is directly tied to confusion around the ambient authority. It's a confusion about whether sockets are supposed to point to synchronization/communication queuing state, or addresses which may be associated with such a state. Considering the anonymous cases, just like memfd on the file side, forces us to get our ontology in order, and consider just the channels themselves, separate from addressability.

----

As a final note, non-userland QUIC (which exists as a broswer-provided, and soon Linux-provided) has something at least superficially similar, which is multiple streams for a given connection. See https://github.com/lxin/quic/pull/53#issuecomment-3309869373 for details, where I am happy to report I was able to advocate for the possibility of giving each stream its own file descriptor.

A sender initiated additional stream + peel off is very similar a connector creating a new connection from a given interface. (Imagine an "interface", "connection", "stream" hierarchy for all 3 concepts.) just as `read` and `write` work on my different types of underlying things, one could imagine `connect` and `accept` *both* working on interfaces (making connections) and connections (making streams).

I don't know all the answers, but I fear that if WASI mostly imitates POSIX, we'll loose the a narrow window of opportunity of ever cleaning up the mess that is the Berkley sockets API, and we'll also loose the opportunity to, by reaping the accidental complexity that Berkley sockets gives us, spend some of the complexity budget without trashing our brains on new things like QUIC sockets.

**Comment by badeend** [comment]
>  I fear that if WASI mostly imitates POSIX, we'll loose the a narrow window of opportunity of ever cleaning up the mess that is the Berkley sockets API

I wholeheartedly agree that the POSIX sockets interface doesn't deserve any beauty awards. But wasi-sockets has very little wiggle room in trying to "fix" it:
- Operating systems expose BSD-style sockets, and
- Guests expect BSD-style sockets

Any additional abstraction that WASI injects in between has very little value in the real world. Because whatever the intermediate abstraction is, the guest (most likely: wasi-libc) will immediately tear down that additional abstraction anyway.

I don't fully comprehend your entire comment, but it sounds like you're after a higher-level networking API that is outside the scope of wasi-sockets.

---

Coming back to the original request in this issue: We can't merge `bind` and `listen` without breaking POSIX compatibility. So closing this one.

**Comment by Ericson2314** [comment]
@badeend By that argument, getting rid of ambient authority was also a mistake, because existing code also expects that.

Having WASI libc rebuild the legacy interface atop something similar still  benefits the ecosystem, because it gives implementations of the wasi syscall interface more flexibility in how they choose to implement sockets. It would be just like how the notion of a current directory is a WASI libc trick.

Also, I fully expect as WASI matures and becomes more common, it's creates an opportunity for language standard libraries to reconsider whether the want to force users to do the legacy song and dance either — the simpler syscalls help incentive simpler programming models down the road.

Conversely, if we just blindly copy the junk that is Berkeley sockets, everyone will shrug, say "good enough", and I doubt this stuff will ever be improved.

**Comment by badeend** [comment]
You're right that code directly targeting WASI p0, p1, p2, p3, p.. will need to be updated after upgrading to a new major version. That's what they signed up for by targeting WASI directly.

However, I expect that to be only a very small amount of use cases. With most consumers going through their language's standard library instead, which in turn likely dispatches to wasi-libc.

> I fully expect as WASI matures and becomes more common, it's creates an opportunity for language standard libraries to reconsider whether the want to force users to do the legacy song and dance either

I can't predict the future nor what impact WASI will have on the API design of existing standard libraries, but I can say this:
POSIX compatibility is a hard requirement for wasi-sockets, and this means that we can't merge `bind` and `listen`. 

As I said before, it sounds like you're after a higher-level networking API that is beyond the scope of wasi-sockets.

**Comment by Ericson2314** [comment]
POSIX only govens what is in the libc, it doesn't govern the underlying system calls.

 If we can make one or the two a no-opt, and just have one underlying system call, that ought to be fully backwards compatible, while allowing language standard libraries that want to to by pass the libc for extra simplicity.
